### PR TITLE
Try running miri on osx

### PR DESCRIPTION
**Pull request by oli-obk**

### CODE REVIEW

**Comment by eddyb** [comment]
The script has:
```sh
#!/bin/bash
```
May I suggest `/bin/sh` or `/usr/bin/env bash`?

**Comment by oli-obk** [comment]
> May I suggest /bin/sh or /usr/bin/env bash?

It works just fine with `bash`, any reason to move? (I have no reason to stay, but "never touch a running system" ;))

**Comment by eddyb** [comment]
@oli-obk `/bin/bash` doesn't work on NixOS, the other two options do, as they are the only portable versions (for some subset of UNIX systems). I was wondering if either of them works on macOS too.

File: .travis.yml [hunk 1]
```diff
@@ -12,7 +17,7 @@ before_script:
 script:
 - |
   # get ourselves a MIR-ful libstd
-  xargo/build.sh
+  cd xargo && RUSTFLAGS='-Zalways-encode-mir' xargo build && cd ..
```
Comment (eddyb) [inline .travis.yml]: Can you keep `xargo/build.sh` here? My point was that *if* the script can be made more portable, it could be used on macOS too.

Comment (oli-obk) [inline .travis.yml]: the portability issues are `readlink`. A quick google search tells me that I need to use `greadlink` after a call to `brew install coreutils`. Which doesn't help portability ;)

**Comment by oli-obk** [comment]
Ok. so we just need to impl `_tlv_atexit`. The c-impl is here: https://opensource.apple.com/source/dyld/dyld-360.22/src/threadLocalVariables.c.auto.html

Looks simple enough, but I'd prefer if someone with a mac could do that.

File: .travis.yml (hunk 1 above)
Comment (RalfJung) [inline .travis.yml]: Wait, macOS has no readlink? Or just doesn't support the flag?
The flag is just needed to make the script work even when called though a symlink. I add a line like this in the beginning of every shell script I write, but we could simplify this one to make it more portable. Just `cd "$(dirname "$0")"` should work.

**Comment by RalfJung** [comment]
Is it not enough to just make that function do nothing? Or does that result in leaks as destructors are not called?

Googling said
> on MacOS, clang uses _tlv_atexit to register destructors for thread_local variables, and then tlv_finalize runs these destructors

Looking at libstd code... I see. On Linux, Rust tries to use `__cxa_thread_atexit_impl` but if that is NULL, it falls back to pthreads TLS. That's the path taken in miri. On macOS, there is no fallback.
So we probably just need some additional state somewhere that records these function-argument pairs and calls them when `tlv_finalize` gets called? (At least that seems to be the usual path. Maybe `tlv_finalize` never gets called for us, as pthreads does this, not Rust.)

All this has me wonder whether we are getting the order right. TLS destructors (and also `_tlv_atexit` if implemented similarly) are called *after* libstd's "at exit"-handlers in miri. We should check whether this is also the case on real systems...

File: .travis.yml (hunk 1 above)
Comment (oli-obk) [inline .travis.yml]: it doesn't support the flag. I'll see if leaving it out will work.

Comment (RalfJung) [inline .travis.yml]: Na that doesn't do the right thing. I wonder which flags it *does* support.

Comment (RalfJung) [inline .travis.yml]: Okay, turns out it supports no flags. So, it's not helpful. I suggest to just use dirname.

**Pull request merged**